spec/ics-23-standard-accumulator/README.md

```diff
@@ -0,0 +1,228 @@
+
+```coffeescript
+createMembershipWitness(AccumulatorState state, Element elem) -> AccumulatorProof
+```
+
+The `createNonMembershipWitness` function generates a proof that an element has not been added to an accumulator.
+
+```coffeescript
+createNonMembersipWitness(AccumulatorState state, Element elem) -> AccumulatorProof
+```
+
+#### Proof verification
+
+The `verifyMembership` function verifies a proof that an element has been added to an accumulator.
+
+```coffeescript
+verifyMembership(AccumulatorRoot root, AccumulatorProof proof, Element elem) -> boolean
+```
+
+The `verifyNonMembership` function verifies a proof that an element has not been added to an accumulator.
+
+```coffeescript
+verifyNonMembership(AccumulatorRoot root, AccumulatorProof proof, Element elem) -> boolean
+```
+
+### Optional functions
+
+An accumulator construction MAY provide the following functions:
+
+The `batchVerifyMembership` function verifies a proof that many elements have been added to an accumulator.
+
+```coffeescript
+batchVerifyMembership(AccumulatorRoot root, AccumulatorProof proof, Set<Element> elems) -> boolean
+```
+
+The `batchVerifyNonMembership` function verifies a proof that many elements have not been added to an accumulator.
+
+```coffeescript
+batchVerifyNonMembership(AccumulatorRoot root, AccumulatorProof proof, Set<Element> elems) -> boolean
+```
+
+If defined, these functions MUST be computationally equivalent to the conjunctive union of `verifyMembership` and `verifyNonMembership` respectively (`proof` may vary):
+
+```coffeescript
+batchVerifyMembership(root, proof, elems) ==
+  verifyMembership(root, proof, elems[0]) &&
+  verifyMembership(root, proof, elems[1]) && ...
+```
+
+```coffeescript
+batchVerifyMembership(root, proof, elems) ==
+  verifyNonMembership(root, proof, elems[0]) &&
+  verifyNonMembership(root, proof, elems[1]) && ...
+```
+
+If batch verification is possible and more efficient than individual verification of one proof per element, an accumulator construction SHOULD define batch verification functions.
+
+### Properties
+
+Accumulators must be *correct* and *sound*. In practice, violations of these properties by computationally-bounded adversaries may be negligible in some security parameter, which is sufficient for use in IBC.
+
+#### Correctness
+
+Accumulator proofs must be *correct*: elements which have been added to the accumulator can always be proved to have been included, but cannot be proved to be excluded.
+
+For an element `elem` in the accumulator `acc`,
+
+```coffeescript
+root = getRoot(acc)
+proof = createMembershipWitness(acc, elem)
+verifyMembership(root, proof, elem) == true
+```
+
+and, likewise, for all values of `proof`,
+
+```coffeescript
+verifyNonMembership(root, proof, elem) == false
+```
+
+#### Soundness
+
+Accumulator proofs must be *sound*: elements which have not been added to the accumulator can never be proved to have been included, but can always be proved to have been excluded.
```

Comment:
Its impossible to construct an accumulator with perfect soundness :P

Reply:
I mention the security parameter above. Do you think it's likely that different chains will want to use the same type of accumulator with different security parameters? (otherwise the security parameter can just be part of the definition of the accumulator algorithm, for the purposes of IBC implementers)

Reply:
I think most accumulatosr should have a fixed security parameter, as you want probably want to do group-specific optimizations in your code. The only accumulator I know of which you can switch security parameters trivially is merkle trees with a hash like blake2b, since the length is just another parameter, and this seems likely to occur imo. (Remember when Cosmos wanted the MT hash size to be 20 bytes)

Reply:
OK :+1: Said as much.